### ISSUE TITLE
How to deal with leap years?

### ISSUE THREAD

**Issue by huaracheguarache** [metno/sea-ice-index-viz#4]
Leap years have an extra day (29th of February) compared to normal years. If we were to plot the data such that the date is included as a part of February all of the non-leap years would have a gap in the data at that point. The current situation in the code is that all days following the 29th of February are "bumped" forward one day in leap years, and we end up with an extra day in December. It appears that NSIDC does the same with its Charctic plotting tool.

Is this the standard way of handling this?

**Comment by TomLav** [comment]
I see that the plotting is using "df.index" (whatever that is) as x:
https://github.com/metno/sea-ice-index-viz/blob/c2e684edfb760ddf32feafdf668bb92de3e54810/bokeh-app/main.py#L72

What is df.index? Is it an integer (index position in the array) or a datetime?

If it is an integer, I would first try to change so that the plotting is indexed on the actual datetime, then see what this results in for leap and non-leap years. But this might be quite a task and then I am not sure this is the first priority.

**Comment by huaracheguarache** [comment]
It looks like it's a type of integer range. This is what happens when I add a `print(df.index)` statement before the line you asked about:

```
RangeIndex(start=0, stop=366, step=1)
```

And this is the result of `print(type(df.index))`:

```
<class 'pandas.core.indexes.range.RangeIndex'>
```

**Comment by TomLav** [comment]
Thank you. Let's keep it in mind for later, that the plotting is based on x as an integer index. I think it would be better to have it based on a datetime, but this is not the highest priority for now.

**Comment by epifanio** [comment]
good catch :) the code used in this plot is the result of one day hack .. I just went trough it and found a first issue:

- xarray converts the dataset into a multi-index data frame which looks like this:

```python
import xarray as xr
nc_url = "https://thredds.met.no/thredds/dodsC/osisaf/met.no/ice/index/v2p1/nh/osisaf_nh_sie_daily.nc"
ds = xr.open_dataset(nc_url)
df = ds.to_dataframe()
df["sie"].loc[df.index.groupby(df.index.get_level_values(0).year)[1979]]
```

this will return the following:
```
time        nv
1979-01-01  0     15.078125
            1     15.078125
1979-01-02  0     15.098750
            1     15.098750
1979-01-03  0     15.119375
                    ...    
1979-12-29  1     14.349687
1979-12-30  0     14.372500
            1     14.372500
1979-12-31  0     14.410000
            1     14.410000
```

Note the `SIE` values are repeated twice for each level found in the `nv` index, and you can see this in the plot as well, the DOY is just wrong .. it is 365*2
 
 To fix it, in `get_data()` I added:
 
 ```python
 df = df.droplevel(1)
 df = df[~df.index.duplicated(keep='first')]
 ```
This removes the multi-index and drops the duplicated time rows. Now at least the DOY looks correct. (I updated the code in the repo with the changes above)


Regarding the index used in the plot, the final data is the folding over 1 Year range of `n`  date-times - to have it indexed with time you will require a real date .. so not sure how to accomplish that.

it is possible to convert the index from integer to  a TimedeltaIndex with:
```python
df.index = pd.timedelta_range(start='1 day', end='366 days', freq='1D')
```
But then the plotting code needs to be adapted to the new data as the `get_ticks()` method will fail.

**Comment by huaracheguarache** [comment]
Here's what's done in the code at this point:

- The data is plotted onto an x-axis with 366 days to accommodate for the extra day from leap years. 
- For the data from individual years that are not leap years there is no data point on the 29th of February. Bokeh will simply linearly interpolate the graph between the 28th of February and 1st of March, and there will be no hovertool value at the 29th for non-leap years.
- The climatological data (min, max, median, percentiles) for the 29th of February is calculated based on the actual data from leap years for the 29th of February and also from linearly interpolated data between the 28th of February and the 1st of March for non-leap years. Using the data from the leap years alone produces quite noticeable discontinuities in the graphs.

**Comment by TomLav** [comment]
That's great. I'll give a look at how the HoverTool behaves.

**Comment by TomLav** [comment]
I looked at the Hover tool. At present it writes the Date (YYYY-MM-DD), the DOY, and the value.
The date and value are ok, but not the DOY. At present, it seems like the DOY is just the index of the point along the x-axis. It means March 1st will always be DOY 61 which is not correct for non-leap years (where DOY should be 60).

Solution: I suggest we remove the indication of DOY. The Date is enough. Also I will open another Issue for the Hover tool to suggest what they could display (in addition to Date and value).

IMO we can close this issue once the DOY is removed from the Hover tool.

**Comment by huaracheguarache** [comment]
Resolved with #29.
